grayscale.py

```diff
@@ -1,336 +1,1055 @@
-        self.gsGraphics1 = thumby.GraphicsClass(self.gsBuffer1, self.width, self.height)
-        self.gsGraphics2 = thumby.GraphicsClass(self.gsBuffer2, self.width, self.height)
+        self.width = 72
+        self.height = 40
+        self.max_x = 72 - 1
+        self.max_y = 40 - 1
 
-        try:
-            self.config = ConfigFile.load()
-        except:
-            self.config = {
-                "displayRefreshTime": 27400
-            }
-            self.saveConfig()
+        self.pages = self.height // 8
+        self.buffer_size = self.pages * self.width
+        self.buffer1 = bytearray(self.buffer_size)
+        self.buffer2 = bytearray(self.buffer_size)
+        self._buffer1 = bytearray(self.buffer_size)
+        self._buffer2 = bytearray(self.buffer_size)
+        self._buffer3 = bytearray(self.buffer_size)
+
+        # The method used to create reduced flicker greyscale using the SSD1306 uses certain
+        # assumptions about the internal behaviour of the controller. Even though the behaviour
+        # seems to back up those assumptions, it is possible that the assumptions are incorrect
+        # but the desired result is achieved anyway. To simplify things, the following comments
+        # are written as if the assumptions _are_ correct.
+
+        # We will keep the display sychronised by resetting the row counter before each frame
+        # and then outputting a frame of 58 rows. This is 18 rows past the 40 of the actual display.
+
+        # Prior to loading in the frame we park the row counter at row 0 and wait for the nominal time
+        # for 9 rows to be output. This (hopefully) provides enough time for the row counter to reach row 0
+        # before it sticks there. (Note: recent test indicate that perhaps the current row actually jumps before parking)
+        # The 'parking' is done by setting the number of rows (aka 'multiplex ratio') to 1 row. This is
+        # an invalid setting according to the datasheet but seems to still have the desired effect.
+        self.pre_frame_cmds = bytearray([0xa8,0, 0xd3,52])
+        # Once the frame has been loaded into the display controller's GDRAM, we set the controller to
+        # output 58 rows, and then delay for the nominal time for 49 rows to be output.
+        # Considering the 18 row 'buffer space' after the real 40 rows, that puts us halfway between the
+        # end of the display, and the row at which it would wrap around.
+        # By having 9 rows either side of the nominal timing, we can absorb any variation in the frequency
+        # of the display controller's RC oscillator as well as any timing offsets introduced by the Python code.
+        self.post_frame_cmds = bytearray([0xd3,40+(64-58), 0xa8,58-1])
+
+        # We enhance the greys by modulating the contrast
+        if False:
+            # brightest
+            self.post_frame_adj = [bytearray([0x81,0x5]), bytearray([0x81,0x7f]), bytearray([0x81,0xff])]
```

Comment:
I'm guessing this is for easy debugging?

Reply:
More just a TBD. I think it looks better using the full contrast range, but that of course doesn't honour the user's config.
Perhaps we could use the config, but with a different set of contrast values.

Reply:
Looking at your comment further down, it will be a bit dim if the maximum contrast value is 127. The controller's contrast command takes values from 0 to 255, but for some reason the thumby.py GraphicsClass's `brightness()` function limits this to 127, and the highest contrast value used in via the config file is 127.
I think I'll make a couple of changes around this so that the full contrast range is used by default, but still using the user's config.